tests/fuzz/s2n_select_server_cert_test.c

```diff
@@ -0,0 +1,176 @@
+#include "tls/s2n_tls_parameters.h"
+#include "utils/s2n_random.h"
+#include "utils/s2n_safety.h"
+#include "s2n_test.h"
+
+#define MAX_TOKENS 256
+
+static void s2n_fuzz_atexit()
+{
+    s2n_cleanup();
+}
+
+int LLVMFuzzerInitialize(const uint8_t *buf, size_t len)
+{
+#ifdef S2N_TEST_IN_FIPS_MODE
+    S2N_TEST_ENTER_FIPS_MODE();
+#endif
+
+    GUARD(s2n_init());
+    GUARD(atexit(s2n_fuzz_atexit));
+    return 0;
+}
+
+/*
+ * Tokenize the input fuzz buffer based on NULL bytes into output array and return the number of tokens.
+ * Avoiding extra heap allocation here to increase fuzz test rate.
+ */
+size_t find_strings(const uint8_t *buf, size_t len, const char **output_strings, size_t max_strings)
+{
+    size_t num_strings = 0;
+    int cursor = 0;
+    while(1) {
+        if (cursor >= len || num_strings == max_strings) {
+            return num_strings;
+        }
+        const char *cur_str = (const char *) (buf + cursor);
+        const char *next_null = (const char *) memchr((const void *) cur_str, '\0', (len - cursor));
+        if (next_null == NULL) {
+            return num_strings;
+        }
+
+        /* We found a null byte. Move the cursor beyond it. */
+        cursor = (((const uint8_t *) next_null - buf) + 1);
+        if (cursor >= len) {
+            return num_strings;
+        }
+        output_strings[num_strings] = cur_str;
+        num_strings++;
+    }
+}
+
+GENERAL_NAME *string_to_general_name(const char *str)
+{
+    ASN1_IA5STRING *asn1_name_str = ASN1_IA5STRING_new();
+    if (!asn1_name_str) {
+        return NULL;
+    }
+
+    if (!ASN1_STRING_set(asn1_name_str, str, strlen(str))) {
+        ASN1_IA5STRING_free(asn1_name_str);
+        return NULL;
+    }
+
+    GENERAL_NAME *san_name = GENERAL_NAME_new();
+    if (!san_name) {
+        ASN1_IA5STRING_free(asn1_name_str);
+        return NULL;
+    }
+
+    GENERAL_NAME_set0_value(san_name, GEN_DNS, asn1_name_str);
+    return san_name;
+}
+
+/*
+ * This fuzz test uses the fuzz input to:
+ * - Generate the data to populate in the SAN of a certificate
+ * - Generate the data to populate the SNI TLS extension
+ * - Fuzz the certificate matching function in s2n: s2n_cert_chain_and_key_matches_name
+ */
+int LLVMFuzzerTestOneInput(const uint8_t *buf, size_t len)
+{
+    struct s2n_cert_chain_and_key *cert = s2n_cert_chain_and_key_new();
+    GENERAL_NAMES* san_names = sk_GENERAL_NAME_new_null();
+    X509 *x509_cert = X509_new();
+
+    if (!x509_cert || !san_names || !cert) {
+        goto cleanup;
+    }
+
+    /* Create an array of strings based on fuzz input. To populate the cert SANs and
+     * input hostname.
+     */
+    const char *strings[MAX_TOKENS] = { NULL };
+    size_t num_strings = find_strings(buf, len, strings, MAX_TOKENS);
+    size_t num_sans;
+    /* This should be a match since we make the hostname the same string as a SAN we've added.
+     * It won't be a match when the SAN is longer than domain name length for SNI.
```

Comment:
How will this case occur? It looks like the string would always match a SAN since we're not passing it through any SNI "filter"?

Reply:
It wouldn't match for cases where the SAN is > 255 bytes since that's the max length we allow for SNI: https://github.com/awslabs/s2n/blob/95e8095a161fef6c9dc4f24bd698acd2a220276e/crypto/s2n_certificate.h#L24

As you mentioned below it's not super important because we aren't check for correctness here.